README.md
# dvi-rs
A parser for the "device independent file" (dvi) format in rust

## TODO
 - At the moment, the "parse" function actually lexes. Could parse into a list
   of pages and fonts.
 - Documentation/examples, currently best examples are the tests


CHANGELOG.md

# 0.2.2

 - Updated `nom` to `5.x`
 - Updated to 2018 edition.

# 0.2.1

 - Corrected Cargo.toml metadata

# 0.2

 - Updated `nom` to `4.x`.


tests/lib.rs
extern crate dvi;
use dvi::Instruction;
use std::fs::File;
use std::io::Read;

fn parse(input: &[u8]) -> Vec<Instruction> {
    let mut input = input;
    let mut instructions = Vec::new();
    while input.len() > 0 {
        let instruction = match Instruction::parse(&input) {
            Result::Ok((i, inst)) => {
                input = i;
                inst
            }
            Result::Err(_) => panic!("Parse error"),
        };
        instructions.push(instruction);
    }
    instructions
}

fn dump(input: &[Instruction]) -> Vec<u8> {
    // will still reallocate but hopefully less
    let mut output = Vec::with_capacity(input.len());
    for inst in input {
        inst.dump(&mut output).unwrap();
    }
    output
}

#[test]
fn main() {
    let mut input_owned = Vec::new();
    File::open("tests/source/main.dvi")
        .unwrap()
        .read_to_end(&mut input_owned)
        .unwrap();
    let instructions = parse(&input_owned);
    // e.g.
    assert!(
        instructions[instructions.len() - 1]
            == Instruction::PostPost {
                post_pointer: 1826,
                ident: 2,
                two_two_three: 5
            }
    );

    let dumped = dump(&instructions);
    // works, but not not guaranteed to in general, since a Vec<Instruciton> has multiple valid
    // dumped representations (because u8 can be stored u32 etc)
    assert_eq!(input_owned, dumped);
    let parsed_again = parse(&dumped);

    for (i, (first, second)) in instructions.iter().zip(parsed_again.iter()).enumerate() {
        assert_eq!(
            first, second,
            "Error: {:?} != {:?}, token no {}",
            first, second, i
        );
    }
    assert_eq!(instructions, parsed_again);
    //println!("{:#?}", instructions);
    //panic!();
}


src/util.rs
use nom::IResult;

/// Get the number of bytes required to store a u32
pub(crate) fn byte_width(val: u32) -> u8 {
    if (val & !0xffu32) == 0 {
        1
    } else if (val & !0xff_ffu32) == 0 {
        2
    } else if (val & !0xff_ff_ffu32) == 0 {
        3
    } else {
        4
    }
}

/// Get the number of bytes required to store an i32
pub(crate) fn byte_width_signed(val: i32) -> u8 {
    // TODO the logic of this could do with looking over - I will remove this message once I can
    // parse documents reliably
    // TODO benchmark against < u8::MAX etc
    if val as u32 & !0x7f == 0 || !val as u32 & !0x7f == 0 {
        1
    } else if val as u32 & !0x7f_ff == 0 || !val as u32 & !0x7f_ff == 0 {
        2
    } else if val as u32 & !0x7f_ff_ff == 0 || !val as u32 & !0x7f_ff_ff == 0 {
        3
    } else {
        4
    }
}

/// A parser to count the number of times the byte 223 occurs. This parser converts Incomplete to
/// Done, which is normally bad, but since these trailing bytes make no difference to the semantic
/// meaning of the document, we don't care if we haven't read them all yet.
pub(crate) fn parse_223(mut i: &[u8]) -> IResult<&[u8], u32, ()> {
    use nom::bytes::streaming::tag;

    let mut count = 0;
    while let Ok((inext, _)) = tag::<_, _, ()>(&[223][..])(i) {
        count += 1;
        i = inext;
    }
    Ok((i, count))
}

#[cfg(test)]
mod tests {

    #[test]
    fn byte_width() {
        use super::byte_width;
        assert_eq!(byte_width(0x7f_ff_ff_ff), 4);
        assert_eq!(byte_width(0x7f_ff_ff), 3);
        assert_eq!(byte_width(0x7f_ff), 2);
        assert_eq!(byte_width(0x7f), 1);
        assert_eq!(byte_width(0), 1);
    }

    #[test]
    fn byte_width_signed() {
        use super::byte_width_signed;
        assert_eq!(byte_width_signed(0), 1);
        assert_eq!(byte_width_signed(0x7f), 1);
        assert_eq!(byte_width_signed(0x7f_ff), 2);
        assert_eq!(byte_width_signed(0x7f_ff_ff), 3);
        assert_eq!(byte_width_signed(0x7f_ff_ff_ff), 4);
        assert_eq!(byte_width_signed(-0x7f), 1);
        assert_eq!(byte_width_signed(-0x7f_ff), 2);
        assert_eq!(byte_width_signed(-0x7f_ff_ff), 3);
        assert_eq!(byte_width_signed(-0x7f_ff_ff_ff), 4);
        assert_eq!(byte_width_signed(::std::i32::MAX), 4);
        assert_eq!(byte_width_signed(::std::i32::MIN), 4);
    }
}


src/parser.rs
//! Parsers for each instruction type

use crate::{util::parse_223, FontDef, Instruction};

use nom::{
    bytes::streaming::take,
    combinator::map,
    number::streaming::{be_i16, be_i24, be_i32, be_i8, be_u16, be_u24, be_u32, be_u8},
    IResult, Needed,
};

pub fn parse(i: &[u8]) -> IResult<&[u8], Instruction, ()> {
    match i.get(0) {
        Some(&op) if op <= 127 => Ok((&i[1..], Instruction::Set(op as u32))),
        Some(&op) if op >= 171 && op <= 234 => Ok((&i[1..], Instruction::Font((op - 171) as u32))),
        Some(&_) => parse_complex(i),
        None => Err(nom::Err::Incomplete(Needed::Unknown)),
    }
}

fn parse_complex(input: &[u8]) -> IResult<&[u8], Instruction, ()> {
    let (input, code) = be_u8(input)?;
    match code {
        // Set
        128 => map(be_u8, |ch| Instruction::Set(ch.into()))(input),
        129 => map(be_u16, |ch| Instruction::Set(ch.into()))(input),
        130 => map(be_u24, Instruction::Set)(input),
        131 => map(be_u32, Instruction::Set)(input),
        // SetRule
        132 => {
            let (input, a) = be_i32(input)?;
            let (input, b) = be_i32(input)?;
            Ok((input, Instruction::SetRule(a, b)))
        }
        // Put
        133 => map(be_u8, |ch| Instruction::Put(ch.into()))(input),
        134 => map(be_u16, |ch| Instruction::Put(ch.into()))(input),
        135 => map(be_u24, Instruction::Put)(input),
        136 => map(be_u32, Instruction::Put)(input),
        // SetRule
        137 => {
            let (input, a) = be_i32(input)?;
            let (input, b) = be_i32(input)?;
            Ok((input, Instruction::PutRule(a, b)))
        }
        // Nop
        138 => Ok((input, Instruction::Nop)),
        // Bop
        139 => {
            let (input, c0) = be_i32(input)?;
            let (input, c1) = be_i32(input)?;
            let (input, c2) = be_i32(input)?;
            let (input, c3) = be_i32(input)?;
            let (input, c4) = be_i32(input)?;
            let (input, c5) = be_i32(input)?;
            let (input, c6) = be_i32(input)?;
            let (input, c7) = be_i32(input)?;
            let (input, c8) = be_i32(input)?;
            let (input, c9) = be_i32(input)?;
            let (input, p) = be_i32(input)?;
            Ok((
                input,
                Instruction::Bop([c0, c1, c2, c3, c4, c5, c6, c7, c8, c9], p),
            ))
        }
        // Eop
        140 => Ok((input, Instruction::Eop)),
        // Push
        141 => Ok((input, Instruction::Push)),
        // Pop
        142 => Ok((input, Instruction::Pop)),
        // Right
        143 => map(be_i8, |a| Instruction::Right(a.into()))(input),
        144 => map(be_i16, |a| Instruction::Right(a.into()))(input),
        145 => map(be_i24, Instruction::Right)(input),
        146 => map(be_i32, Instruction::Right)(input),
        // W
        147 => Ok((input, Instruction::W(None))),
        148 => map(be_i8, |a| Instruction::W(Some(a.into())))(input),
        149 => map(be_i16, |a| Instruction::W(Some(a.into())))(input),
        150 => map(be_i24, |a| Instruction::W(Some(a)))(input),
        151 => map(be_i32, |a| Instruction::W(Some(a)))(input),
        // X
        152 => Ok((input, Instruction::X(None))),
        153 => map(be_i8, |a| Instruction::X(Some(a.into())))(input),
        154 => map(be_i16, |a| Instruction::X(Some(a.into())))(input),
        155 => map(be_i24, |a| Instruction::X(Some(a)))(input),
        156 => map(be_i32, |a| Instruction::X(Some(a)))(input),
        // Down
        157 => map(be_i8, |a| Instruction::Down(a.into()))(input),
        158 => map(be_i16, |a| Instruction::Down(a.into()))(input),
        159 => map(be_i24, Instruction::Down)(input),
        160 => map(be_i32, Instruction::Down)(input),
        // Y
        161 => Ok((input, Instruction::Y(None))),
        162 => map(be_i8, |a| Instruction::Y(Some(a.into())))(input),
        163 => map(be_i16, |a| Instruction::Y(Some(a.into())))(input),
        164 => map(be_i24, |a| Instruction::Y(Some(a)))(input),
        165 => map(be_i32, |a| Instruction::Y(Some(a)))(input),
        // Z
        166 => Ok((input, Instruction::Z(None))),
        167 => map(be_i8, |a| Instruction::Z(Some(a.into())))(input),
        168 => map(be_i16, |a| Instruction::Z(Some(a.into())))(input),
        169 => map(be_i24, |a| Instruction::Z(Some(a)))(input),
        170 => map(be_i32, |a| Instruction::Z(Some(a)))(input),
        // Font
        235 => map(be_u8, |f| Instruction::Font(f.into()))(input),
        236 => map(be_u16, |f| Instruction::Font(f.into()))(input),
        237 => map(be_u24, Instruction::Font)(input),
        238 => map(be_u32, Instruction::Font)(input),
        // Xxx
        239 => {
            let (input, length) = be_u8(input)?;
            let (input, slice) = take(length)(input)?;
            Ok((input, Instruction::Xxx(slice.to_owned())))
        }
        240 => {
            let (input, length) = be_u16(input)?;
            let (input, slice) = take(length)(input)?;
            Ok((input, Instruction::Xxx(slice.to_owned())))
        }
        241 => {
            let (input, length) = be_u24(input)?;
            let (input, slice) = take(length)(input)?;
            Ok((input, Instruction::Xxx(slice.to_owned())))
        }
        242 => {
            let (input, length) = be_u32(input)?;
            let (input, slice) = take(length)(input)?;
            Ok((input, Instruction::Xxx(slice.to_owned())))
        }
        // FontDef
        243 => {
            let (input, number) = be_u8(input)?;
            font_def(input, number.into())
        }
        244 => {
            let (input, number) = be_u16(input)?;
            font_def(input, number.into())
        }
        245 => {
            let (input, number) = be_u24(input)?;
            font_def(input, number)
        }
        246 => {
            let (input, number) = be_u32(input)?;
            font_def(input, number)
        }
        // Pre
        247 => {
            let (input, format) = be_u8(input)?;
            let (input, numerator) = be_u32(input)?;
            let (input, denominator) = be_u32(input)?;
            let (input, magnification) = be_u32(input)?;
            let (input, comment_length) = be_u8(input)?;
            let (input, comment) = take(comment_length)(input)?;
            Ok((
                input,
                Instruction::Pre {
                    format,
                    numerator,
                    denominator,
                    magnification,
                    comment: comment.to_owned(),
                },
            ))
        }
        248 => {
            let (input, final_bop_pointer) = be_i32(input)?;
            let (input, numerator) = be_u32(input)?;
            let (input, denominator) = be_u32(input)?;
            let (input, magnification) = be_u32(input)?;
            let (input, tallest_height) = be_i32(input)?;
            let (input, widest_width) = be_i32(input)?;
            let (input, max_stack_depth) = be_u16(input)?;
            let (input, total_no_pages) = be_u16(input)?;
            Ok((
                input,
                Instruction::Post {
                    final_bop_pointer,
                    numerator,
                    denominator,
                    magnification,
                    tallest_height,
                    widest_width,
                    max_stack_depth,
                    total_no_pages,
                },
            ))
        }
        249 => {
            let (input, post_pointer) = be_u32(input)?;
            let (input, ident) = be_u8(input)?;
            let (input, two_two_three) = parse_223(input)?;
            Ok((
                input,
                Instruction::PostPost {
                    post_pointer,
                    ident,
                    two_two_three,
                },
            ))
        }
        _ => unreachable!(),
    }
}

fn font_def(input: &[u8], number: u32) -> IResult<&[u8], Instruction, ()> {
    let (input, checksum) = be_u32(input)?;
    let (input, scale_factor) = be_u32(input)?;
    let (input, design_size) = be_u32(input)?;
    let (input, directory_len) = be_u8(input)?;
    let (input, filename_len) = be_u8(input)?;
    let (input, directory) = take(directory_len)(input)?;
    let (input, filename) = take(filename_len)(input)?;
    let directory = match directory_len {
        0 => None,
        _ => Some(directory.to_owned()),
    };
    Ok((
        input,
        Instruction::FontDef(FontDef {
            number,
            checksum,
            scale_factor,
            design_size,
            directory,
            filename: filename.to_owned(),
        }),
    ))
}


examples/tester.rs
use dvi::{FontDef, Instruction};
use std::{fs, path::PathBuf};
use structopt::StructOpt;

type Result<T = (), E = anyhow::Error> = std::result::Result<T, E>;

#[derive(Debug, StructOpt)]
struct Opt {
    #[structopt(parse(from_os_str))]
    file: PathBuf,
}

fn main() -> Result {
    let opt = Opt::from_args();
    let file = fs::read(&opt.file)?;
    Dvi::new(InstIter(&file));
    Ok(())
}

pub struct Dvi<I> {
    instr: I,
    scale: f64,
}

impl<I> Dvi<I>
where
    I: Iterator<Item = Result<Instruction, &'static str>>,
{
    fn new(mut input: I) -> Result<Self, &'static str> {
        let pre = input.next().ok_or("empty file")??;
        let scale = if let Instruction::Pre {
            format,
            numerator,
            denominator,
            magnification,
            comment,
        } = pre
        {
            if format != 2 {
                return Err("unsupported format");
            }
            eprintln!("{}", String::from_utf8_lossy(&comment));
            (numerator as f64 * magnification as f64) / (denominator as f64 * 1000.0)
        } else {
            return Err("first instruction was not a pre");
        };
        Ok(Dvi {
            instr: input,
            scale,
        })
    }
}

struct InstIter<'a>(&'a [u8]);

impl<'a> Iterator for InstIter<'a> {
    type Item = Result<Instruction, &'static str>;
    fn next(&mut self) -> Option<Self::Item> {
        if self.0.is_empty() {
            return None;
        }
        let (input, inst) = match Instruction::parse(self.0) {
            Ok((i, inst)) => (i, inst),
            Err(_) => return Some(Err("malformed instruction")),
        };
        self.0 = input;
        Some(Ok(inst))
    }
}


src/dumper.rs
//! functions to dump correct output to stream for each instruction type
use crate::{
    util::{byte_width, byte_width_signed},
    FontDef, Instruction,
};
use byteorder::{BigEndian, WriteBytesExt};
use std::io::{self, Write};

// Helper macros

/// Write out an i32 or u32, using smallest space possible
///
/// This is just purely a space-saving macro. The 5 number variant uses an Option<x32> and the 4
/// number variant uses x32
macro_rules! write_small {
    ( signed $v:expr , $writer:ident => $code1:expr , $code2:expr ,
      $code3:expr , $code4:expr ) => {{
        // help the type checker
        let out: io::Result<()> = {
            match byte_width_signed($v) {
                1 => {
                    $writer.write_u8($code1)?;
                    $writer.write_i8($v as i8)?;
                }
                2 => {
                    $writer.write_u8($code2)?;
                    $writer.write_i16::<BigEndian>($v as i16)?;
                }
                3 => {
                    $writer.write_u8($code3)?;
                    $writer.write_i24::<BigEndian>($v)?;
                }
                4 => {
                    $writer.write_u8($code4)?;
                    $writer.write_i32::<BigEndian>($v)?;
                }
                _ => { unreachable!() }
            };
            Ok(())
        };
        out
    }};

    ( signed $v:expr , $writer:ident => $code0:expr , $code1:expr ,
      $code2:expr , $code3:expr , $code4:expr ) => {{
        match $v {
            Some(v) => write_small!(signed v, $writer => $code1, $code2, $code3, $code4)?,
            None => $writer.write_u8($code0)?
        };
        Ok(())
    }};

    ( unsigned $v:expr , $writer:ident => $code1:expr ,
      $code2:expr , $code3:expr , $code4:expr ) => {{
        let out: io::Result<()> = {
            match byte_width($v) {
                1 => {
                    $writer.write_u8($code1)?;
                    $writer.write_u8($v as u8)?;
                }
                2 => {
                    $writer.write_u8($code2)?;
                    $writer.write_u16::<BigEndian>($v as u16)?;
                }
                3 => {
                    $writer.write_u8($code3)?;
                    $writer.write_u24::<BigEndian>($v)?;
                }
                4 => {
                    $writer.write_u8($code4)?;
                    $writer.write_u32::<BigEndian>($v)?;
                }
                _ => { unreachable!() }
            };
            Ok(())
        };
        out
    }};

    ( unsigned $v:expr , $writer:ident => $code0:expr , $code1:expr , $code2:expr ,
      $code3:expr , $code4:expr ) => {{
        match $v {
            Some(v) => write_small!(unsigned v, $writer => $code1, $code2, $code3, $code4),
            None => $writer.write_u8($code0)?
        };
        Ok(())
    }};
}

// Encoders

/// Dump an instruction to an impl of Write
pub(crate) fn dump<W: Write>(i: &Instruction, writer: &mut W) -> io::Result<()> {
    match *i {
        Instruction::Set(ch) => dump_set(ch, writer),
        Instruction::SetRule(a, b) => dump_set_rule(a, b, writer),
        Instruction::Put(ch) => dump_put(ch, writer),
        Instruction::PutRule(a, b) => dump_put_rule(a, b, writer),
        Instruction::Nop => writer.write_u8(138),
        Instruction::Bop(c, p) => dump_bop(c, p, writer),
        Instruction::Eop => writer.write_u8(140),
        Instruction::Push => writer.write_u8(141),
        Instruction::Pop => writer.write_u8(142),
        Instruction::Right(amt) => dump_right(amt, writer),
        Instruction::W(amt) => dump_w(amt, writer),
        Instruction::X(amt) => dump_x(amt, writer),
        Instruction::Down(amt) => dump_down(amt, writer),
        Instruction::Y(amt) => dump_y(amt, writer),
        Instruction::Z(amt) => dump_z(amt, writer),
        Instruction::Font(num) => dump_font(num, writer),
        Instruction::Xxx(ref data) => dump_xxx(&data[..], writer),
        Instruction::FontDef(ref font_def) => dump_font_def(font_def, writer),
        Instruction::Pre {
            format,
            numerator,
            denominator,
            magnification,
            ref comment,
        } => dump_pre(
            format,
            numerator,
            denominator,
            magnification,
            comment,
            writer,
        ),
        Instruction::Post {
            final_bop_pointer,
            numerator,
            denominator,
            magnification,
            tallest_height,
            widest_width,
            max_stack_depth,
            total_no_pages,
        } => dump_post(
            final_bop_pointer,
            numerator,
            denominator,
            magnification,
            tallest_height,
            widest_width,
            max_stack_depth,
            total_no_pages,
            writer,
        ),

        Instruction::PostPost {
            post_pointer,
            ident,
            two_two_three,
        } => dump_postpost(post_pointer, ident, two_two_three, writer),
    }
}

fn dump_set<W: Write>(ch: u32, writer: &mut W) -> io::Result<()> {
    match byte_width(ch) {
        1 => {
            if ch < 128 {
                writer.write_u8(ch as u8)?;
            } else {
                writer.write_u8(128)?;
                writer.write_u8(ch as u8)?;
            }
        }
        2 => {
            writer.write_u8(129)?;
            writer.write_u16::<BigEndian>(ch as u16)?;
        }
        3 => {
            writer.write_u8(130)?;
            writer.write_u24::<BigEndian>(ch)?;
        }
        4 => {
            writer.write_u8(131)?;
            writer.write_u32::<BigEndian>(ch)?;
        }
        _ => unreachable!(),
    };
    Ok(())
}

fn dump_set_rule<W: Write>(a: i32, b: i32, writer: &mut W) -> io::Result<()> {
    writer.write_u8(132)?;
    writer.write_i32::<BigEndian>(a)?;
    writer.write_i32::<BigEndian>(b)?;
    Ok(())
}

fn dump_put<W: Write>(v: u32, writer: &mut W) -> io::Result<()> {
    write_small!(unsigned v, writer => 133, 134, 135, 136)
}

fn dump_put_rule<W: Write>(a: i32, b: i32, writer: &mut W) -> io::Result<()> {
    writer.write_u8(137)?;
    writer.write_i32::<BigEndian>(a)?;
    writer.write_i32::<BigEndian>(b)?;
    Ok(())
}

fn dump_bop<W: Write>(c: [i32; 10], p: i32, writer: &mut W) -> io::Result<()> {
    writer.write_u8(139)?;
    for ci in c.iter() {
        writer.write_i32::<BigEndian>(*ci)?;
    }
    writer.write_i32::<BigEndian>(p)?;
    Ok(())
}

fn dump_right<W: Write>(v: i32, writer: &mut W) -> io::Result<()> {
    write_small!(signed v, writer => 143, 144, 145, 146)
}

fn dump_w<W: Write>(v: Option<i32>, writer: &mut W) -> io::Result<()> {
    write_small!(signed v, writer => 147, 148, 149, 150, 151)
}

fn dump_x<W: Write>(v: Option<i32>, writer: &mut W) -> io::Result<()> {
    write_small!(signed v, writer => 152, 153, 154, 155, 156)
}

fn dump_down<W: Write>(v: i32, writer: &mut W) -> io::Result<()> {
    write_small!(signed v, writer => 157, 158, 159, 160)
}

fn dump_y<W: Write>(v: Option<i32>, writer: &mut W) -> io::Result<()> {
    write_small!(signed v, writer => 161, 162, 163, 164, 165)
}

fn dump_z<W: Write>(v: Option<i32>, writer: &mut W) -> io::Result<()> {
    write_small!(signed v, writer => 166, 167, 168, 169, 170)
}

fn dump_font<W: Write>(f: u32, writer: &mut W) -> io::Result<()> {
    match byte_width(f) {
        1 => {
            if f <= 63 {
                writer.write_u8((f + 171) as u8)?;
            } else {
                writer.write_u8(235)?;
                writer.write_u8(f as u8)?;
            }
        }
        2 => {
            writer.write_u8(236)?;
            writer.write_u16::<BigEndian>(f as u16)?;
        }
        3 => {
            writer.write_u8(237)?;
            writer.write_u24::<BigEndian>(f)?;
        }
        4 => {
            writer.write_u8(238)?;
            writer.write_u32::<BigEndian>(f)?;
        }
        _ => unreachable!(),
    };
    Ok(())
}

fn dump_xxx<W: Write>(data: &[u8], writer: &mut W) -> io::Result<()> {
    assert!(
        data.len() < ::std::u32::MAX as usize,
        "The length of extention data won't fit in 32 bits"
    );
    write_small!(unsigned data.len() as u32, writer => 239, 240, 241, 242)?;
    writer.write_all(data)?;
    Ok(())
}

/// Helper for `dump_font_def`
#[inline]
fn dump_font_def_helper<W: Write>(v: u32, writer: &mut W) -> io::Result<()> {
    write_small!(unsigned v, writer => 243, 244, 245, 246)
}

fn dump_font_def<W: Write>(def: &FontDef, writer: &mut W) -> io::Result<()> {
    assert!(
        def.filename.len() <= ::std::u8::MAX as usize,
        "Filename too long in Font Definition"
    );
    assert!(
        if let Some(ref d) = def.directory {
            d.len() <= ::std::u8::MAX as usize
        } else {
            true
        },
        "Directory name too long in Font Definition"
    );
    dump_font_def_helper(def.number, writer)?;
    writer.write_u32::<BigEndian>(def.checksum)?;
    writer.write_u32::<BigEndian>(def.scale_factor)?;
    writer.write_u32::<BigEndian>(def.design_size)?;
    match def.directory {
        Some(ref d) => writer.write_u8(d.len() as u8),
        None => writer.write_u8(0),
    }?;
    writer.write_u8(def.filename.len() as u8)?;
    if let Some(ref d) = def.directory {
        writer.write_all(&d[..])?
    };
    writer.write_all(&def.filename[..])?;
    Ok(())
}

/// Pre
fn dump_pre<W: Write>(
    format: u8,
    numerator: u32,
    denominator: u32,
    magnification: u32,
    comment: &[u8],
    writer: &mut W,
) -> io::Result<()> {
    assert!(comment.len() < 0x100, "Comment length must fit into u8");
    writer.write_u8(247)?;
    writer.write_u8(format)?;
    writer.write_u32::<BigEndian>(numerator)?;
    writer.write_u32::<BigEndian>(denominator)?;
    writer.write_u32::<BigEndian>(magnification)?;
    writer.write_u8(comment.len() as u8)?;
    writer.write_all(comment)?;
    Ok(())
}

/// Post
fn dump_post<W: Write>(
    final_bop_pointer: i32,
    numerator: u32,
    denominator: u32,
    magnification: u32,
    tallest_height: i32,
    widest_width: i32,
    max_stack_depth: u16,
    total_no_pages: u16,
    writer: &mut W,
) -> io::Result<()> {
    writer.write_u8(248)?;
    writer.write_i32::<BigEndian>(final_bop_pointer)?;
    writer.write_u32::<BigEndian>(numerator)?;
    writer.write_u32::<BigEndian>(denominator)?;
    writer.write_u32::<BigEndian>(magnification)?;
    writer.write_i32::<BigEndian>(tallest_height)?;
    writer.write_i32::<BigEndian>(widest_width)?;
    writer.write_u16::<BigEndian>(max_stack_depth)?;
    writer.write_u16::<BigEndian>(total_no_pages)?;
    Ok(())
}

fn dump_postpost<W: Write>(
    post_pointer: u32,
    ident: u8,
    two_two_three: u32,
    writer: &mut W,
) -> io::Result<()> {
    writer.write_u8(249)?;
    writer.write_u32::<BigEndian>(post_pointer)?;
    writer.write_u8(ident)?;
    for _ in 0..two_two_three {
        writer.write_u8(223)?;
    }
    Ok(())
}
